p2p/state.py

```diff
@@ -101,74 +101,95 @@ def register_peer(self, peer: BasePeer) -> None:
-                # Slow down if we've reached the limit of pending nodes.
-                self.logger.debug("Pending trie nodes limit reached, sleeping a bit")
-                await asyncio.sleep(0.3)
+            if len(self._pending_nodes) < max_pending:
+                requests = self.scheduler.next_batch(eth.MAX_STATE_FETCH)
+                if not requests:
+                    # Although we frequently yield control above, to let our msg handler process
+                    # received nodes (scheduling new requests), there may be cases when the
+                    # pending nodes take a while to arrive thus causing the scheduler to run out
+                    # of new requests for a while.
+                    self.logger.debug("Scheduler queue is empty, sleeping a bit")
+                    await wait_with_token(asyncio.sleep(0.5), token=self.cancel_token)
+                else:
+                    await self.request_nodes([request.node_key for request in requests])
             else:
-                # Yield control to ensure the Peer's msg_handler callback is called to process any
-                # nodes we may have received already. Otherwise we spin too fast and don't process
-                # received nodes often enough.
-                await asyncio.sleep(0)
-
-            self._maybe_report_progress()
+                self.logger.debug("Pending trie nodes limit reached, sleeping a bit")
+                await wait_with_token(asyncio.sleep(0.5), token=self.cancel_token)
 
         self.logger.info("Finished state sync with root hash %s", encode_hex(self.root_hash))
 
-    def _maybe_report_progress(self):
-        if (time.time() - self._last_report_time) >= self._report_interval:
-            self._last_report_time = time.time()
+    async def _periodically_report_progress(self):
+        while True:
+            now = time.time()
+            self.logger.info("====== State sync progress ========")
             self.logger.info("Nodes processed: %d", self._total_processed_nodes)
+            self.logger.info("Nodes processed per second (average): %d",
+                             self._total_processed_nodes / (now - self._start_time))
+            self.logger.info("Nodes committed to DB: %d", self.scheduler.committed_nodes)
             self.logger.info(
                 "Nodes requested but not received yet: %d", len(self._pending_nodes))
             self.logger.info(
                 "Nodes scheduled but not requested yet: %d", len(self.scheduler.requests))
+            self.logger.info("Total nodes timed out: %d", self._total_timeouts)
+            try:
+                await wait_with_token(asyncio.sleep(self._report_interval), token=self.cancel_token)
+            except OperationCancelled:
+                break
 
 
 class StateSync(HexaryTrieSync):
 
+    def __init__(self, root_hash, db):
+        super().__init__(root_hash, db, logging.getLogger("p2p.state.StateSync"))
```

Comment:
Note to self:

Might be nice to drop the `logger` from `__init__` at some point and just have it as a class property.  Subclasses could easily override with their own logger.